### PR TITLE
#167730738 Build signup feature with token as response

### PR DESCRIPTION
**Pull request by zanio**
#### What does this PR do?
It Builds signup feature and provide token as response and also send email to newly signed up user for account verification. When User verifies his/her email the user table field ( `isVerified`) is updated to true
### Description of Task to be completed?
N/A
#### How should this be manually tested?
**Please Note:** This is a follow up ticket to the ticket [ft-email-verification-167730740](https://github.com/andela/firestar-backend/tree/ft-email-verification-167730740)

1. ```git clone -b ft-signup-user-167730738  https://github.com/andela/firestar-backend.git .```  
2. run `npm install` and create  a `.env file` and provide the following information to it 
```
DB_USERNAME=database username
DB_PASSWORD=database password
DB_NAME=name of database for development
DB_HOST=Host url
PORT=port number to listen for connections
DB_TEST_NAME=name of test database
JWT_SECRET=secret variable to sign and decode tokens
SECRET_KEY_SIGN_UP = choose any random character
EMAIL_VERIFY_TOKEN_SECRET_KEY = choose any random character
TOKENEXPIRY=18000
CLIENT_URL="https://firestar_backend.herokuapp.com"
YOUR_EMAIL="" (description: email to create test account and receive reset email link, must be able to access the email)

```
Please kindly make a reference to the file `env.example` for their description and the slack channel for this project for any `env` variable value

3. Create two databases name in your postgres database ( `barefoot and barefoot_test` )

image

4. Run `npm run migrate:dev` to create a user table. ** Note At any point in time you can drop this table by runing `sequelize db:migrate:undo:all` and run `npm run seed:dev` to seed some datas. The purpose for this is to make sure our roles table is filled with sample data

5. start the server with `npm run start:dev`

6. Now use postman and go to `localhost:PORT/api/v1/users/auth/register` using a `POST` request with a body of 
```
body:{ 
email: 'youremail',//A valid email address, you would be sent verification link
firstName: 'Yourfirstname', // At least 3 characters long and less than 30
lastName: 'Yourlastname', // At least 3 characters long and less than 30
password: 'Strongpassword' // Password must be at least 8 character long, with at least an uppercase, lowercase, digit and special character
}
```
Screenshot (58)

Also feel free to use invalid details to see the different response on each error

7. If successful you would see a response containing the token similar to the response of the screenshot on postman or the below response, also check your database to be sure it was added and `isVerified` is `false`
```
{
    "status": "success",
    "message": "user Added!",
    "data": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NCwiaWF0esoxNTY3NTUzMzg2fQ.JQe7CEY4okMlsQafEa3TSzqQ0-IpgdFq1v1NN75E8bg",
        "emailToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6Im5vZGVkZXdldd5YWhvby5jb20iLCJpYXQiOjE1Njc1NTMzNzgsImV4cCI6MTU2NzYzOTc3OH0.20Pk4DbKGH_-quxsm26gLczauaeAZ7dxrMd3lzaQ044",
        "id": 4,
        "email": "nodedeweb@yahoo.com",
        "firstName": "Aniefiok",
        "lastName": "Akpan"
    }
}
```
Screenshot (59)

**Observe as well that the the user table contain a field `isVerified` which checks if user is verified.**

8. Visit the email you provided and see the newly delivered email which contains your name and link to verify your account. Make sure your server is still ruining. **Click on this email or copy full URL address**
Screenshot (60)

9. make a `get` request with the copied link that looks like this `http://localhost:PORT/api/v1/users/email/verify?id=token` to verify your account

Screenshot (61)

10.  Use any GUI database tool to see your User table `isVerified` updated from `false` to `true` which indicates that your account has been verified. Also you can see the newly inserted data from logins table

12. You can also decide to access other protected routes. to do this copy the token gotten during signup and use it in postman header as `Authorization`. An example of a protected route is `http://localhost:PORT/api/v1/users/myaccount`  You can test it as well

Screenshot (63)

13. To start using this feature on your own ticket all you have to do is call in two middleware named `authorization` and `jwtVerify` as shown on the screenshot

Screenshot (64)

#### Any background context you want to provide?
Please note that two token are used, In this feature.
**token that is used as query string on verification link which also expires in 24hrs**
**User token upon signup which would be used as Authorization in header to access other protected route**

#### What are the relevant pivotal tracker stories?
[#167730738](https://www.pivotaltracker.com/story/show/167730738)

#### Screenshots or gifs (if appropriate)
**Test coverage information**

image

* 

#### Questions:
* N/A
Great job @zanio 

correct `EMAIL_VERIFY_TOKEN_SECRET_KEY` to `SECRET_KEY_EMAIL_VERIFY_TOKEN`  in
`firestar-backend/src/utils/index.js`

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        id, email, firstName, lastName,
+      } = await userService.addUser(user);
+      const token = await jwtSignUser(id);
+      util.setSuccess(201, 'user Added!', {
+        token, emailToken, id, email, firstName, lastName,
+      });
+      return util.send(res);
+    } catch (error) {
+      if (error.original.routine === '_bt_check_unique') {
+        util.setError(400, 'Email already exist');
```
Comment (OlawaleJoseph) [inline src/controllers/userController.js]: Why use 400 status code? This should be 409.

File: src/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      return util.send(res);
+    }
+  }
+
+  /**
+ * @param {req} req that contains the req body object.
+ * @param {res} res content to be rendered.
+ * @returns {object} Success or failure response on getting a specific user
+ */
+  static async getAUser(req, res) {
```
Comment (OlawaleJoseph) [inline src/controllers/userController.js]: Why this? Shouldn't you just be creating users?

Comment (OlawaleJoseph) [inline src/controllers/userController.js]: ditto

File: src/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      return util.send(res);
+    }
+  }
+
+  /**
+ * @param {req} req that contains the req body object.
+ * @param {res} res content to be rendered.
+ * @returns {object} Success or failure response to delete a specific user.
+ */
+  static async deleteUser(req, res) {
```
Comment (OlawaleJoseph) [inline src/controllers/userController.js]: ditto

File: src/middlewares/auth/auth.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    req.token = token;
+    next();
+  } else {
+    res.status(401).json({ status: 401, error: 'You are not authorized to access this route' });
+  }
+};
+
+export const jwtVerify = (req, res, next) => {
+  const { token } = req;
+  jwt.verify(token, process.env.SECRET_KEY_SIGN_UP, (err, result) => {
```
Comment (OlawaleJoseph) [inline src/middlewares/auth/auth.js:20]: Why secret key_sign_up? Are there different secrets for other routes? You can just say JWT_SECRET

File: src/middlewares/auth/auth.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    if (err) {
+      res.status(401).json({ status: 401, error: 'failed jwt verification' });
```
Comment (OlawaleJoseph) [inline src/middlewares/auth/auth.js]: No reason as to why the JWT verification failed. Use a more suitable error message.

File: src/middlewares/mail.js [hunk 6]
```diff
@@ -14,14 +15,16 @@ import Mail from '../services/mail/Mail';
  */
 export const SendVerificationEmail = async (req, res, next) => {
   let { email, firstName, lastName } = req.body;
+  console.log(email)
```
Comment (OlawaleJoseph) [inline src/middlewares/mail.js]: Please remove this

File: src/middlewares/mail.js [hunk 7]
```diff
@@ -14,14 +15,16 @@ import Mail from '../services/mail/Mail';
...
   email = email ? email.trim() : '';
   firstName = firstName ? firstName.trim() : '';
   lastName = lastName ? lastName.trim() : '';
   /**
    * @var {id} id is the user unique id from Table column
    */
-  const id = 'some_encoded_identiity';
+  const id = req.user ? req.user.email : 'some_encoded_identity';
```
Comment (OlawaleJoseph) [inline src/middlewares/mail.js]: What do you mean by some encoded entity?

File: src/middlewares/mail.js [hunk 8]
```diff
@@ -48,7 +50,10 @@ export const SendVerificationEmail = async (req, res, next) => {
       return res.status(403).json({ status: 403, error: response.message });
     }
     if (response.message === 'queryA ECONNREFUSED smtp.gmail.com') {
-      return res.status(409).json({ status: 409, error: response.message });
+      return res.status(409).json({ status: 409, error: 'Network error occured, please check your network' });
```
Comment (OlawaleJoseph) [inline src/middlewares/mail.js]: Kindly use a different status code. 409 status code is for conflict not network related issues

File: src/middlewares/validation/validation.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/* eslint-disable camelcase */
```
Comment (OlawaleJoseph) [inline src/middlewares/validation/validation.js]: why disable camelCase?

File: src/middlewares/validation/validation.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    email, password, firstName, lastName
+  };
+  try {
+    const { error, value } = validateData(signupBody, signUpValidationSchema);
+    const errMessage = error ? error.details[0].message : null;
+    if (error) {
+      util.setError(400, errMessage);
+      return util.send(res);
+    }
+    if (error) {
```
Comment (OlawaleJoseph) [inline src/middlewares/validation/validation.js]: What is the difference between this and you have on line 18?

File: src/seeders/20190815005541-create-users.js [hunk 11]
```diff
@@ -4,15 +4,17 @@ const faker = require('faker');
 
 module.exports = {
   up: (queryInterface, Sequelize) => {
-    return queryInterface.bulkInsert('Users', [{
+    return queryInterface.bulkInsert('user', [{
       first_name: faker.name.firstName(),
```
Comment (OlawaleJoseph) [inline src/seeders/20190815005541-create-users.js]: Use camelCase

File: src/services/userservice.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
+/* eslint-disable no-useless-catch */
+import db from '../models/index';
+/**
+ * @param { class } provide response to user signup activity.
+ */
+class userService {
+  /**
+ * @returns {object} of all query data from databas
+ */
+  static async getAllUser() {
```
Comment (OlawaleJoseph) [inline src/services/userservice.js]: Why did you create this for sign up?

File: src/services/userservice.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+      throw error;
+    }
+  }
+
+  /**
+   * @param { id } id of user to be updated.
+   * @param { updateUser } updateUser of user to be updated.
+ * @returns {object} object containing recently details of user.
+ */
+  static async updateUser(id, updateUser) {
```
Comment (OlawaleJoseph) [inline src/services/userservice.js]: Why this in sign up?

File: src/services/userservice.js [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+      throw error;
+    }
+  }
+
+  /**
+   * @param { email } email of user to be updated.
+   * @param { updateUser } updateUser of user to be updated.
+ * @returns {object} object containing recently details of user.
+ */
+  static async updateUserByEmail(email, updateUser) {
```
Comment (OlawaleJoseph) [inline src/services/userservice.js:45]: ditto

File: src/services/userservice.js [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+   * @param { id } id of user to delete.
+ * @returns {object} object of user deleted from the database.
+ */
+  static async deleteUser(id) {
```
Comment (OlawaleJoseph) [inline src/services/userservice.js]: ditto

File: src/middlewares/validation/validation.js (hunk 9 above)
Comment (OlawaleJoseph) [reply src/middlewares/validation/validation.js]: According to what the TTL told me. Move the validation folder into the helpers folder.

@OlawaleJoseph This is a middleware, it cannot be there. it is a validation middleware

**Review by OlawaleJoseph — CHANGES_REQUESTED** [review]
Good work

**Review by ola357 — CHANGES_REQUESTED** [review]
1. Please give an implementation plan or workflow, how the user gets to signup
@ola357 itwould be done on the slack group

2. please let the controller and routes names represent their responsibilities
@ola357 yes it has been done

3. why are you protecting the index or welcome route, please the example in the comments

@ola357  i did that so in case anyone want to know how to implement or make use of the the middleware. For example you would need to pass `authorization` and `jwtVerify` if you want to access a protected route

4. Can you shift the hashing password helper function to the User model, let the User model take care of hashing a password it wants to save
Good job by the way

File: src/controllers/userController.js (hunk 2 above)
Comment (zanio) [reply src/controllers/userController.js]: Well i wanted to handle all CRUD operation of user. But nevertheless it has been deleted

File: src/controllers/userController.js (hunk 1 above)
Comment (zanio) [reply src/controllers/userController.js]: Alright

File: src/middlewares/auth/auth.js (hunk 4 above)
Comment (zanio) [reply src/middlewares/auth/auth.js:20]: Yes they are different key for different routes, for example. For example the secret key to generate token sent for email verification is different for the secret key when a user signs up and token is giving. So we have two tokens. signup token and email token

File: src/middlewares/auth/auth.js (hunk 5 above)
Comment (zanio) [reply src/middlewares/auth/auth.js]: Done

File: src/middlewares/mail.js (hunk 6 above)
Comment (zanio) [reply src/middlewares/mail.js]: Done

File: src/middlewares/mail.js (hunk 7 above)
Comment (zanio) [reply src/middlewares/mail.js]: It has been fixed

File: src/middlewares/validation/validation.js (hunk 9 above)
Comment (zanio) [reply src/middlewares/validation/validation.js]: Alright, it has be done

File: src/middlewares/validation/validation.js (hunk 10 above)
Comment (zanio) [reply src/middlewares/validation/validation.js]: That was a mistake

File: src/seeders/20190815005541-create-users.js (hunk 11 above)
Comment (zanio) [reply src/seeders/20190815005541-create-users.js]: Yes, it has be done

File: src/services/userservice.js (hunk 12 above)
Comment (zanio) [reply src/services/userservice.js]: it has been deleted

File: src/middlewares/mail.js (hunk 8 above)
Comment (zanio) [reply src/middlewares/mail.js]: done

File: src/middlewares/auth/auth.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+import dotenv from 'dotenv';
+
+dotenv.config();
+
+export const authorization = (req, res, next) => {
+  const header = req.header('Authorization');
+
+  if (typeof header !== 'undefined') {
+    const bearer = header;
+    const token = bearer;
```
Comment (OlawaleJoseph) [inline src/middlewares/auth/auth.js]: Why assign bearer to header then token to bearer?

File: src/utils/validation/emailValidation.js [hunk 17]
```diff
@@ -0,0 +1,8 @@
+/* eslint-disable import/prefer-default-export */
```
Comment (OlawaleJoseph) [inline src/utils/validation/emailValidation.js:1]: Don't disable `import/prefer-default-export`

File: src/utils/index.js [hunk 18]
```diff
@@ -7,12 +7,24 @@ dotenv.config();
...
- * @param {element} element of that array
- * @return {integer} index of the  element in array
+ * @param {id} id of newly registered user from the user Table.
+ * @return {string} generated token that is a string data type.
  */
-export const arrayTest = (array, element) => array.indexOf(element);
+export const jwtSignUser = (id) => new Promise((resolve, reject) => {
+  const token = jwt.sign({ id }, process.env.SECRET_KEY_SIGN_UP);
+  if (token) return resolve(token);
+  const ERROR = 'Token cannot be assigned';
```
Comment (OlawaleJoseph) [inline src/utils/index.js]: A more descriptive error as to why the token cannot be verified. e.g. Invalid Token

**Comment by OlawaleJoseph** [comment]
I can use numbers as first name and last name. Please ensure only alphabets are allowed
Screenshot (39)

File: src/middlewares/auth/auth.js (hunk 16 above)
Comment (zanio) [reply src/middlewares/auth/auth.js]: I extracted the token first from the header by splitting it to bearer and token

File: src/utils/validation/emailValidation.js (hunk 17 above)
Comment (zanio) [reply src/utils/validation/emailValidation.js:1]: I have to disable it

File: src/utils/index.js (hunk 18 above)
Comment (zanio) [reply src/utils/index.js]: Aiit

**Comment by zanio** [comment]
> I can use numbers as first name and last name. Please ensure only alphabets are allowed
> Screenshot (39)

@OlawaleJoseph I have fixed that

File: src/controllers/userController.js [hunk 19]
```diff
@@ -0,0 +1,42 @@
+/* eslint-disable camelcase */
+import userService from '../services/userservice';
```
Comment (marusoft) [inline src/controllers/userController.js:6]: please use camelcase i.e '../services/userService'; then remove th comment

File: src/services/userservice.js [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+class userService {
+  /**
+   * @param { newUser } newUser to be added to database.
+ * @returns {object} containing newly added user to the database
+ */
+  static async addUser(newUser) {
+    try {
+      return await db.users.create(newUser);
+    } catch (error) {
+      throw error;
```
Comment (marusoft) [inline src/services/userservice.js:15]: Good work, you can replace `throw error` with 'return error.message;' in order to get rid of this comment `/* eslint-disable no-useless-catch */`

**Review by marusoft — CHANGES_REQUESTED** [review]

File: src/controllers/userController.js (hunk 19 above)
Comment (zanio) [reply src/controllers/userController.js:6]: okay

File: src/services/userservice.js (hunk 20 above)
Comment (zanio) [reply src/services/userservice.js:15]: Thanks. Fixed

File: src/utils/index.js [hunk 21]
```diff
@@ -7,12 +7,24 @@ dotenv.config();
  * @return {string} generated token that is a string data type.
  */
 export const emailVerifyToken = (id) => new Promise((resolve) => {
-  const token = jwt.sign({ id }, process.env.SECRET_KEY_EMAIL_VERIFY_TOKEN, { expiresIn: '24h' });
+  const token = jwt.sign({ id }, process.env.EMAIL_VERIFY_TOKEN_SECRET_KEY, { expiresIn: '24h' });
```
Comment (marusoft) [inline src/utils/index.js:12]: This is conflicting with what you have in your .env.example
correct `EMAIL_VERIFY_TOKEN_SECRET_KEY` to `SECRET_KEY_EMAIL_VERIFY_TOKEN` in
`firestar-backend/src/utils/index.js`

**Review by marusoft — CHANGES_REQUESTED** [review]
Please correct the request changes

**Review by OlawaleJoseph — APPROVED** [review]
Great work

File: src/utils/index.js (hunk 21 above)
Comment (zanio) [reply src/utils/index.js:12]: I have two tokens. token for email generation and token for accessing protected route. I used different secret keys to generate and verify token

**Review by ola357 — APPROVED** [review]

**Review by marusoft — APPROVED** [review]
Awesome @zanio

**Pull request merged**